### ISSUE TITLE
Text rendering should attempt to respect locale (OpenType locl) variants for characters

### ISSUE THREAD

**Issue by Artoria2e5** [mapnik/mapnik#3655]
Quite a few Unified CJK Ideographs ("Chinese Characters") are written in different ways in different countries/languages. Such difference is why Adobe and Google has released a bunch of local variants for their Source Han Sans/Noto Sans CJK fonts. In Mapnik for example, at https://www.openstreetmap.org/way/25097203, the "门" (door, gate) character (the last char on its own line) happens to be presented in the Japanese form with a vertical stroke over its rectangular body:

![image](https://cloud.githubusercontent.com/assets/6459309/24877043/e20d59dc-1dfb-11e7-94fe-3102fe4b7c3f.png)

Since these locations are found in China, a Chinese form is expected for these characters. Such form is shown below, from a screenshot of, well, this issue:
![image](https://cloud.githubusercontent.com/assets/6459309/24877055/eb01ed82-1dfb-11e7-8154-8c6eba8a1c6c.png)

(There are quite a few other "门"s to the north.)

I understand that Mapnik, generally only taking the `name` part without a language label attached, cannot reliably infer the language used for the string and therefore the `locl` value to use for font rendering (as well as just the language value to feed into stuff like fontconfig). But with a crude heuristic based on national/regional borders, Mapnik should be able to get most of the cases right.

A Latin-script (Romanian) example of such a `locl` feature is available at https://en.wikipedia.org/wiki/Pango#Support_for_OpenType_features. Also try pasting "关门吃草逛街" into the four previews at https://typekit.com/lists/source-han-sans.

**Comment by springmeyer** [comment]
Hi @Artoria2e5 - thanks for the thoughtful recommendations. If I am understanding right you are referring the the "Mapnik" style on osm.org rather than specifically the Mapnik codebase. This is the bug tracker for the Mapnik codebase. I think you should create this issue over at https://github.com/gravitystorm/openstreetmap-carto to discuss with the style authors.

**Comment by Artoria2e5** [comment]
I believe the stylesheet won't have any say over what locale to use unless mapnik figures out what language to use.

**Comment by tomhughes** [comment]
@Artoria2e5 I don't see any reason why the stylesheet can't do that with a suitable SQL query, though it would probably be more efficient for osm2pgsql to do it while importing the data.

**Comment by Artoria2e5** [comment]
Cross-posted the issue there. Thanks.

**Comment by sommerluk** [comment]
@springmeyer 

Does mapnik offer any possiblity to control the language used for the _locl_ OpenType feature?

Apparently the _locl_ feature itself can be enabled or disabled with _font-feature-settings_. However, if  _locl_ is enabled, the engine is supposed to use localized glyphs, therefor the text passed to the engine needs to be escorted by the information in which language the text is written. _font-feature-settings_ is based on CSS syntax. In CSS, there is no need to specify the language of text, because this information is stored in the HTML file that contains the text.

How can I transmit this language information to Mapnik – not globally, but individually for each label? Can I?

**Comment by springmeyer** [comment]
@sommerluk I don't know. What I can say is that the `font-feature-settings` feature in mapnik is implemented by passing the args to https://github.com/behdad/harfbuzz, the library used in Mapnik to shape text. You can see the settings being passed [here](https://github.com/mapnik/mapnik/blob/9e7c842a26ae70aaf1710bc4bc3a111ce87cf6f6/include/mapnik/text/harfbuzz_shaper.hpp#L124). It looks like harfbuzz has a `hb_buffer_set_language` API that  Mapnik does not use. Perhaps that is related.

**Comment by sommerluk** [comment]
@springmeyer Thanks! That was the code I was searching for. So, indeed _locl_ can be enabled and disabled in Mapnik by font-feature-settings, but for actually get localized rendering, Mapnik would have to implement something to access hb_buffer_set_language. http://harfbuzz.freedesktop.narkive.com/1l03EBMh/how-to-use-locl-feature-of-noto-sans-cjk-with-harfbuzz gives some more details about HarfBuzz’ language API. Do you think there are chances to implement this in further versions of Mapnik?

**Comment by springmeyer** [comment]
@sommerluk yes, this looks like something that could be implemented. It looks like what needs to happen is:

 - Start using the `hb_buffer_set_language` api
 - Add an option to the Mapnik `<TextSymbolizer>` for `language` (could also be called `locale` or `locl`) that would accept a string that is assumed to represent a valid `ISO 639` language code.
 - That property would be data-driven meaning it could be pulled from the database

Once implemented in Mapnik then you could apply this regionally by pre-calculating a field in the database for the value for the `language` to pass to mapnik. Then you would author a stylesheet with a `<TextSymbolizer>` like:

```xml
        <Rule>
            <TextSymbolizer face-name="Noto Sans ...." language="[locale]" "font-feature-settings="[features]">[text]</TextSymbolizer>
        </Rule>
```

Where `locale` would be the pre-calculated `ISO 639` value field and `features` would be the pre-calculated (optionaly) value that would include the necessary syntax enabling `locl`.

A sample patch that implements this based on the system locale you can play around with is:

```diff
diff --git a/include/mapnik/text/harfbuzz_shaper.hpp b/include/mapnik/text/harfbuzz_shaper.hpp
index bee1195..c6203b0 100644
--- a/include/mapnik/text/harfbuzz_shaper.hpp
+++ b/include/mapnik/text/harfbuzz_shaper.hpp
@@ -114,6 +114,8 @@ static void shape_text(text_line & line,
             hb_buffer_add_utf16(buffer.get(), uchar_to_utf16(text.getBuffer()), text.length(), text_item.start, static_cast<int>(text_item.end - text_item.start));
             hb_buffer_set_direction(buffer.get(), (text_item.dir == UBIDI_RTL) ? HB_DIRECTION_RTL : HB_DIRECTION_LTR);
             hb_buffer_set_script(buffer.get(), _icu_script_to_script(text_item.script));
+            // TODO: accept ISO 639 language code via stylesheet?
+            hb_buffer_set_language(buffer.get(), hb_language_get_default());
             hb_font_t *font(hb_ft_font_create(face->get_face(), nullptr));
             // https://github.com/mapnik/test-data-visual/pull/25
```

**Comment by artemp** [comment]
@sommerluk @springmeyer 
We can `guestimate` language from script. This won't be 100% accurate but good enough I think. Also this is a good candidate for a default:   

```c++
static hb_language_t script_to_language(hb_script_t script)
{
    switch (script) {
        // Unicode 1.1
        case HB_SCRIPT_ARABIC: return hb_language_from_string("ar", -1); break;
        case HB_SCRIPT_ARMENIAN: return hb_language_from_string("hy", -1); break;
        case HB_SCRIPT_BENGALI: return hb_language_from_string("bn", -1); break;
        case HB_SCRIPT_CANADIAN_ABORIGINAL: return hb_language_from_string("iu", -1); break;
        case HB_SCRIPT_CHEROKEE: return hb_language_from_string("chr", -1); break;
        case HB_SCRIPT_COPTIC: return hb_language_from_string("cop", -1); break;
        case HB_SCRIPT_CYRILLIC: return hb_language_from_string("ru", -1); break;
        case HB_SCRIPT_DEVANAGARI: return hb_language_from_string("hi", -1); break;
        case HB_SCRIPT_GEORGIAN: return hb_language_from_string("ka", -1); break;
        case HB_SCRIPT_GREEK: return hb_language_from_string("el", -1); break;
        case HB_SCRIPT_GUJARATI: return hb_language_from_string("gu", -1); break;
        case HB_SCRIPT_GURMUKHI: return hb_language_from_string("pa", -1); break;
        case HB_SCRIPT_HANGUL: return hb_language_from_string("ko", -1); break;
        case HB_SCRIPT_HEBREW: return hb_language_from_string("he", -1); break;
        case HB_SCRIPT_HIRAGANA: return hb_language_from_string("ja", -1); break;
        case HB_SCRIPT_KANNADA: return hb_language_from_string("kn", -1); break;
        case HB_SCRIPT_KATAKANA: return hb_language_from_string("ja", -1); break;
        case HB_SCRIPT_LAO: return hb_language_from_string("lo", -1); break;
        case HB_SCRIPT_LATIN: return hb_language_from_string("en", -1); break;
        case HB_SCRIPT_MALAYALAM: return hb_language_from_string("ml", -1); break;
        case HB_SCRIPT_MONGOLIAN: return hb_language_from_string("mn", -1); break;
        case HB_SCRIPT_ORIYA: return hb_language_from_string("or", -1); break;
        case HB_SCRIPT_SYRIAC: return hb_language_from_string("syr", -1); break;
        case HB_SCRIPT_TAMIL: return hb_language_from_string("ta", -1); break;
        case HB_SCRIPT_TELUGU: return hb_language_from_string("te", -1); break;
        case HB_SCRIPT_THAI: return hb_language_from_string("th", -1); break;

        // Unicode 2.0
        case HB_SCRIPT_TIBETAN: return hb_language_from_string("bo", -1); break;

        // Unicode 3.0
        case HB_SCRIPT_ETHIOPIC: return hb_language_from_string("am", -1); break;
        case HB_SCRIPT_KHMER: return hb_language_from_string("km", -1); break;
        case HB_SCRIPT_MYANMAR: return hb_language_from_string("my", -1); break;
        case HB_SCRIPT_SINHALA: return hb_language_from_string("si", -1); break;
        case HB_SCRIPT_THAANA: return hb_language_from_string("dv", -1); break;

        // Unicode 3.2
        case HB_SCRIPT_BUHID: return hb_language_from_string("bku", -1); break;
        case HB_SCRIPT_HANUNOO: return hb_language_from_string("hnn", -1); break;
        case HB_SCRIPT_TAGALOG: return hb_language_from_string("tl", -1); break;
        case HB_SCRIPT_TAGBANWA: return hb_language_from_string("tbw", -1); break;

        // Unicode 4.0
        case HB_SCRIPT_UGARITIC: return hb_language_from_string("uga", -1); break;

        // Unicode 4.1
        case HB_SCRIPT_BUGINESE: return hb_language_from_string("bug", -1); break;
        case HB_SCRIPT_OLD_PERSIAN: return hb_language_from_string("peo", -1); break;
        case HB_SCRIPT_SYLOTI_NAGRI: return hb_language_from_string("syl", -1); break;

        // Unicode 5.0
        case HB_SCRIPT_NKO: return hb_language_from_string("nko", -1); break;

        // no representative language exists
        default: return HB_LANGUAGE_INVALID; break;
    }
}
```   

And later : 

```c++
.....
if (language != HB_LANGUAGE_INVALID)
{
    hb_buffer_set_language(buffer.get(), script_to_language(script)); // set most common language for the run based script            
}

````

**Comment by sommerluk** [comment]
@artemp Yes, I think `script_to_language` is the right way to get a default/fallback value for the language.

This would not be the same for a hole Mapnik style, but individually for each label, right?

Note that for solving the Han unification issue, it is furthermore necessary to be able to override the default value in the stylesheet.

**Comment by springmeyer** [comment]
> Note that for solving the Han unification issue, it is furthermore necessary to be able to override the default value in the stylesheet.

From skimming https://github.com/gravitystorm/openstreetmap-carto/issues/2208 this was my understanding too. Perhaps an expression via the db could be supported and if it evaluated to null then we'd default with `script_to_language`. This would allow SQL queries to solve the Han unification problem with custom logic (maybe regional results) and then Mapnik to default to something reasonable if nothing is provided.

**Comment by artemp** [comment]
> This would not be the same for a hole Mapnik style, but individually for each label, right?
This would per text `run` in each individual label. 

/cc @sommerluk

**Comment by artemp** [comment]
85aebf72688bb3715df77c46fe17fd722249e142 adds support for `guessing` language from script.


```xml
<TextSymbolizer fontset-name="Noto" fill="red" size="32"  placement="point" dx="-15" dy="-10">
      "天安门"
    </TextSymbolizer>
```
The above text classified as `Hani` by harfbuzz and the result: 

<img width="108" alt="hani" src="https://cloud.githubusercontent.com/assets/59056/25653311/a34ec246-2fec-11e7-8ee4-2ccfa6eed280.png">

/cc @sommerluk @springmeyer @Artoria2e5

**Comment by sommerluk** [comment]
@artemp Thanks a lot! That looks great!

Adding a parameter like _language=*_ to TextSymbolizer and ShieldSymbolizer would be the only thing that remains, right?

**Comment by artemp** [comment]
@sommerluk - yes, I think so ^

**Comment by sommerluk** [comment]
For the record: If I understand it correctly, HarfBuzz does not necessarily require the language tags to be ISO 639, which would provide language information only. HarfBuzz seems to accepts [all valid BCP-47 tags](http://harfbuzz.freedesktop.narkive.com/1l03EBMh/how-to-use-locl-feature-of-noto-sans-cjk-with-harfbuzz#post5), which is great because this way you cannot only provide the language, but also the script and the region. And that’s the only way to differentiate the various variants of Chinese writing. If in the future Mapnik will pass this value directly to HarfBuzz, I guess this means that also Mapnik will support all BCP-47 tags, right?

**Comment by kocio-pl** [comment]
Hi, what are current chances of adding missing `language=*` option? Are there any plans or obstacles?
